fe/fe-core/src/main/java/com/starrocks/mysql/privilege/AuthUpgrader.java

```diff
@@ -0,0 +1,433 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.mysql.privilege;
+
+import com.starrocks.analysis.ResourcePattern;
+import com.starrocks.analysis.TablePattern;
+import com.starrocks.analysis.UserIdentity;
+import com.starrocks.authentication.AuthenticationException;
+import com.starrocks.authentication.AuthenticationManager;
+import com.starrocks.common.Config;
+import com.starrocks.privilege.ActionSet;
+import com.starrocks.privilege.PEntryObject;
+import com.starrocks.privilege.PrivilegeCollection;
+import com.starrocks.privilege.PrivilegeException;
+import com.starrocks.privilege.PrivilegeManager;
+import com.starrocks.privilege.PrivilegeTypes;
+import com.starrocks.privilege.RolePrivilegeCollection;
+import com.starrocks.privilege.UserPrivilegeCollection;
+import com.starrocks.server.GlobalStateMgr;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class AuthUpgrader {
+
+    private static final Logger LOG = LogManager.getLogger(AuthUpgrader.class);
+    private Auth auth;
+    private AuthenticationManager authenticationManager;
+    private PrivilegeManager privilegeManager;
+    private GlobalStateMgr globalStateMgr;
+
+    // constants used when upgrading
+    private short tableTypeId;
+    private short userTypeId;
+    private String tableTypeStr;
+    private String dbTypeStr;
+    private String userTypeStr;
+    private List<PEntryObject> allTablesInAllDbObject;
+    private ActionSet selectActionSet;
+    private ActionSet impersonateActionSet;
+
+    public AuthUpgrader(
+            Auth auth,
+            AuthenticationManager authenticationManager,
+            PrivilegeManager privilegeManager,
+            GlobalStateMgr globalStateMgr) {
+        this.auth = auth;
+        this.authenticationManager = authenticationManager;
+        this.privilegeManager = privilegeManager;
+        this.globalStateMgr = globalStateMgr;
+    }
+
+    public boolean needUpgrade() {
+        return !this.authenticationManager.isLoaded() || !this.privilegeManager.isLoaded();
+    }
+
+    public void upgradeAsLeader() throws RuntimeException {
+        try {
+            LOG.info("start to upgrade as leader.");
+            init();
+            upgradeUser();
+            Map<String, Long> roleNameToId = upgradeRole(null);
+            this.globalStateMgr.getEditLog().logAuthUpgrade(roleNameToId);
+            LOG.info("upgraded as leader successfully.");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void replayUpgradeAsFollower(Map<String, Long> roleNameToId) throws AuthUpgradeUnrecoveredException {
```

Comment:
AsFollower is not accurate, because leader will also call this function when restarted.

Reply:
That's totally my mistake. I'll fix it. Thanks a lot!